File: package/typescript/src/hitAndBlow.ts
/**
 * Hit & Blow の判定を行う関数
 *
 * @param answer - 正解の数値のリスト
 * @param guess - プレイヤーの推測リスト (同じ長さ)
 * @returns [Hit数, Blow数]
 */
export function hitAndBlow(answer: number[], guess: number[]): [number, number] {
  // Count hits (same position and same value)
  const hits = answer.reduce((count, val, i) => count + (val === guess[i] ? 1 : 0), 0);

  // Count total matches (regardless of position)
  const answerCounter = new Map<number, number>();
  const guessCounter = new Map<number, number>();

  // Count occurrences in answer
  for (const num of answer) {
    answerCounter.set(num, (answerCounter.get(num) || 0) + 1);
  }

  // Count occurrences in guess
  for (const num of guess) {
    guessCounter.set(num, (guessCounter.get(num) || 0) + 1);
  }

  // Total matches is the sum of minimum counts for each number
  let totalMatches = 0;
  for (const [num, answerCount] of answerCounter) {
    const guessCount = guessCounter.get(num) || 0;
    totalMatches += Math.min(answerCount, guessCount);
  }

  // Blows = total matches - hits
  const blows = totalMatches - hits;

  return [hits, blows];
}

File: package/python/src/hit_and_blow.py
from collections import Counter


def hit_and_blow(answer: list[int], guess: list[int]) -> tuple[int, int]:
    """
    Hit & Blow の判定を行う関数

    Args:
        answer (List[int]): 正解の数値のリスト
        guess (List[int]): プレイヤーの推測リスト (同じ長さ)

    Returns:
        Tuple[int, int]: (Hit数, Blow数)
    """
    # Count hits (same position and same value)
    hits = sum(1 for i in range(len(answer)) if answer[i] == guess[i])

    # Count total matches (regardless of position)
    answer_counter = Counter(answer)
    guess_counter = Counter(guess)

    # Total matches is the sum of minimum counts for each number
    total_matches = sum(min(answer_counter[num], guess_counter[num]) for num in answer_counter)

    # Blows = total matches - hits
    blows = total_matches - hits

    return (hits, blows)
